cache/refs.go

```diff
@@ -627,109 +624,116 @@ func (sr *immutableRef) ociDesc(ctx context.Context, dhs DescHandlers) (ocispecs
-	compressionVariantAnnotationsLabelPrefix = "buildkit.io/compression/annotation."
-	compressionVariantMediaTypeLabel         = "buildkit.io/compression/mediatype"
+	blobVariantGCLabel         = "containerd.io/gc.ref.content.blob-"
+	blobAnnotationsLabelPrefix = "buildkit.io/blob/annotation."
+	blobMediaTypeLabel         = "buildkit.io/blob/mediatype"
 )
 
-func compressionVariantDigestLabel(compressionType compression.Type) string {
-	return compressionVariantDigestLabelPrefix + compressionType.String()
-}
-
-func getCompressionVariants(ctx context.Context, cs content.Store, dgst digest.Digest) (res []compression.Type, _ error) {
-	info, err := cs.Info(ctx, dgst)
-	if errors.Is(err, errdefs.ErrNotFound) {
-		return nil, nil
-	} else if err != nil {
-		return nil, err
-	}
-	for k := range info.Labels {
-		if strings.HasPrefix(k, compressionVariantDigestLabelPrefix) {
-			if t := compression.Parse(strings.TrimPrefix(k, compressionVariantDigestLabelPrefix)); t != compression.UnknownCompression {
-				res = append(res, t)
-			}
-		}
-	}
-	return
-}
-
-func (sr *immutableRef) getCompressionBlob(ctx context.Context, compressionType compression.Type) (ocispecs.Descriptor, error) {
-	return getCompressionVariantBlob(ctx, sr.cm.ContentStore, sr.getBlob(), compressionType)
-}
-
-func getCompressionVariantBlob(ctx context.Context, cs content.Store, dgst digest.Digest, compressionType compression.Type) (ocispecs.Descriptor, error) {
-	info, err := cs.Info(ctx, dgst)
-	if err != nil {
-		return ocispecs.Descriptor{}, err
-	}
-	dgstS, ok := info.Labels[compressionVariantDigestLabel(compressionType)]
-	if ok {
-		dgst, err := digest.Parse(dgstS)
-		if err != nil {
-			return ocispecs.Descriptor{}, err
-		}
-		return getBlobDesc(ctx, cs, dgst)
-	}
-	return ocispecs.Descriptor{}, errdefs.ErrNotFound
-}
-
-func (sr *immutableRef) addCompressionBlob(ctx context.Context, desc ocispecs.Descriptor, compressionType compression.Type) error {
+func (sr *immutableRef) addBlob(ctx context.Context, desc ocispecs.Descriptor) error {
 	cs := sr.cm.ContentStore
-	if err := sr.cm.LeaseManager.AddResource(ctx, leases.Lease{ID: sr.ID()}, leases.Resource{
-		ID:   desc.Digest.String(),
-		Type: "content",
-	}); err != nil {
+	blobDigest := sr.getBlob()
+	info, err := cs.Info(ctx, blobDigest)
+	if err != nil {
 		return err
 	}
-	info, err := cs.Info(ctx, sr.getBlob())
+	vInfo, err := cs.Info(ctx, desc.Digest)
 	if err != nil {
 		return err
 	}
-	if info.Labels == nil {
-		info.Labels = make(map[string]string)
+	vInfo.Labels = map[string]string{
+		blobVariantGCLabel + blobDigest.String(): blobDigest.String(),
 	}
-	cachedVariantLabel := compressionVariantDigestLabel(compressionType)
-	info.Labels[cachedVariantLabel] = desc.Digest.String()
-	if _, err := cs.Update(ctx, info, "labels."+cachedVariantLabel); err != nil {
+	vInfo = addBlobDescToInfo(desc, vInfo)
+	if _, err := cs.Update(ctx, vInfo, fieldsFromLabels(vInfo.Labels)...); err != nil {
```

Comment:
Doesn't this create loops that are never cleaned up because they keep references to each other?

Reply:
Thank you for the comment. Contents are cleaned up even if a loop of GC label is created when no lease reference them. Added a test to make sure that.